// @tanstack/nitro-v2-vite-plugin/dist/cjs/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
let nitropack = require("nitropack");
let pathe = require("pathe");
//#region src/index.ts
var ssrBundle;
var ssrEntryFile;
function isFullUrl(str) {
	try {
		new URL(str);
		return true;
	} catch {
		return false;
	}
}
function nitroV2Plugin(nitroConfig) {
	let resolvedConfig;
	return [{
		name: "tanstack-nitro-v2-vite-plugin",
		generateBundle: { handler(_options, bundle) {
			if (this.environment.name !== "ssr") return;
			let entryFile;
			for (const [_name, file] of Object.entries(bundle)) if (file.type === "chunk") {
				if (file.isEntry) {
					if (entryFile !== void 0) this.error(`Multiple entry points found for service "${this.environment.name}". Only one entry point is allowed.`);
					entryFile = file.fileName;
				}
			}
			if (entryFile === void 0) this.error(`No entry point found for service "${this.environment.name}".`);
			ssrEntryFile = entryFile;
			ssrBundle = bundle;
		} },
		configResolved(config) {
			resolvedConfig = config;
		},
		config(_, env) {
			if (env.command !== "build") return;
			return {
				environments: { ssr: {
					consumer: "server",
					build: {
						ssr: true,
						write: false,
						copyPublicDir: false,
						commonjsOptions: { include: [/node_modules/] }
					}
				} },
				builder: {
					sharedPlugins: true,
					async buildApp(builder) {
						const client = builder.environments.client;
						const server = builder.environments.ssr;
						if (!client) throw new Error("Client environment not found");
						if (!server) throw new Error("SSR environment not found");
						await builder.build(client);
						await builder.build(server);
						const virtualEntry = "#tanstack/start/entry";
						const nitro = await (0, nitropack.createNitro)({
							baseURL: !isFullUrl(resolvedConfig.base) ? resolvedConfig.base : void 0,
							publicAssets: [{
								dir: client.config.build.outDir,
								maxAge: 31536e3,
								baseURL: "/"
							}],
							...nitroConfig,
							esbuild: { options: {
								target: server.config.build.target || void 0,
								...nitroConfig?.esbuild?.options
							} },
							renderer: virtualEntry,
							rollupConfig: {
								...nitroConfig?.rollupConfig,
								plugins: [virtualBundlePlugin(ssrBundle)]
							},
							virtual: {
								...nitroConfig?.virtual,
								[virtualEntry]: `import { fromWebHandler } from 'h3'
                                    import handler from '${ssrEntryFile}' 
                                    export default fromWebHandler(handler.fetch)`
							}
						});
						await (0, nitropack.prepare)(nitro);
						await (0, nitropack.copyPublicAssets)(nitro);
						await (0, nitropack.build)(nitro);
						await nitro.close();
					}
				}
			};
		}
	}];
}
function virtualBundlePlugin(bundle) {
	let _modules = null;
	const getModules = () => {
		if (_modules) return _modules;
		_modules = /* @__PURE__ */ new Map();
		for (const [fileName, content] of Object.entries(bundle)) if (content.type === "chunk") {
			const virtualModule = {
				code: content.code,
				map: null
			};
			const maybeMap = bundle[`${fileName}.map`];
			if (maybeMap && maybeMap.type === "asset") virtualModule.map = maybeMap.source;
			_modules.set(fileName, virtualModule);
			_modules.set((0, pathe.resolve)(fileName), virtualModule);
		}
		return _modules;
	};
	return {
		name: "virtual-bundle",
		resolveId(id, importer) {
			const modules = getModules();
			if (modules.has(id)) return (0, pathe.resolve)(id);
			if (importer) {
				const resolved = (0, pathe.resolve)((0, pathe.dirname)(importer), id);
				if (modules.has(resolved)) return resolved;
			}
			return null;
		},
		load(id) {
			const m = getModules().get(id);
			if (!m) return null;
			return m;
		}
	};
}
//#endregion
exports.nitroV2Plugin = nitroV2Plugin;
